frame/contracts/src/exec.rs

```diff
@@ -378,456 +396,640 @@ pub trait Executable<T: Config>: Sized {
-		if self.depth == 0 {
-			debug_assert!(self.self_trie_id.is_none());
-			debug_assert!(self.caller.is_none());
-			debug_assert!(ContractInfoOf::<T>::get(&self.self_account).is_none());
-			TransactorKind::PlainAccount
-		} else {
-			TransactorKind::Contract
-		}
+	/// Reference to the current (top) frame.
+	fn frame(&self) -> &Frame<T> {
+		self.frames.last().unwrap_or(&self.first_frame)
 	}
-}
 
-/// Describes whether we deal with a contract or a plain account.
-enum TransactorKind {
-	/// Transaction was initiated from a plain account. That can be either be through a
-	/// signed transaction or through RPC.
-	PlainAccount,
-	/// The call was initiated by a contract account.
-	Contract,
-}
+	/// Mutable reference to the current (top) frame.
+	fn frame_mut(&mut self) -> &mut Frame<T> {
+		self.frames.last_mut().unwrap_or(&mut self.first_frame)
+	}
 
-/// Describes possible transfer causes.
-enum TransferCause {
-	Call,
-	Instantiate,
-	Terminate,
-}
+	/// Iterator over all frames.
+	///
+	/// The iterator starts with the top frame and ends with the root frame.
+	fn frames(&self) -> impl Iterator<Item=&Frame<T>> {
+		sp_std::iter::once(&self.first_frame)
+			.chain(&self.frames)
+			.rev()
+	}
 
-/// Transfer some funds from `transactor` to `dest`.
-///
-/// We only allow allow for draining all funds of the sender if `cause` is
-/// is specified as `Terminate`. Otherwise, any transfer that would bring the sender below the
-/// subsistence threshold (for contracts) or the existential deposit (for plain accounts)
-/// results in an error.
-fn transfer<T: Config>(
-	cause: TransferCause,
-	origin: TransactorKind,
-	transactor: &T::AccountId,
-	dest: &T::AccountId,
-	value: BalanceOf<T>,
-) -> DispatchResult
-where
-	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
-{
-	use self::TransferCause::*;
-	use self::TransactorKind::*;
-
-	// Only seal_terminate is allowed to bring the sender below the subsistence
-	// threshold or even existential deposit.
-	let existence_requirement = match (cause, origin) {
-		(Terminate, _) => ExistenceRequirement::AllowDeath,
-		(_, Contract) => {
-			ensure!(
-				T::Currency::total_balance(transactor).saturating_sub(value) >=
-					Contracts::<T>::subsistence_threshold(),
-				Error::<T>::BelowSubsistenceThreshold,
-			);
-			ExistenceRequirement::KeepAlive
-		},
-		(_, PlainAccount) => ExistenceRequirement::KeepAlive,
-	};
+	/// Same as `frames` but with a mutable reference as iterator item.
+	fn frames_mut(&mut self) -> impl Iterator<Item=&mut Frame<T>> {
+		sp_std::iter::once(&mut self.first_frame)
+			.chain(&mut self.frames)
+			.rev()
+	}
 
-	T::Currency::transfer(transactor, dest, value, existence_requirement)
-		.map_err(|_| Error::<T>::TransferFailed)?;
+	/// Returns whether the current contract is on the stack multiple times.
+	fn is_recursive(&self) -> bool {
+		let account_id = &self.frame().account_id;
+		self.frames().skip(1).any(|f| &f.account_id == account_id)
```

Comment:
why do we skip the first frame here ?

Reply:
since it is iterated reversed, it makes sense to me now

Reply:
The first frame is actually the top frame which is the current frame. So if we don't skip it the expression would be always true.